eip-0020.md

```diff
@@ -0,0 +1,246 @@
+transaction should satisfy that condition by adding required context variables,
+creating expected number of outputs with specific registers depending on the
+contract. There is no way for a universal wallet application to know all
+the specific details of all the possible contracts.
+
+On the contrary, every Ergo dApp (think of website) is usually built on top of
+specific contracts. Business logic of the dApp includes spending of the boxes
+protected by those application-specific contracts.
+In other words, any dApp is in full control of creating transactions and thus
+can spend its own contracts without above mentioned problems. However, the dApp
+cannot sign those transactions, because signing requires knowledge of private keys
+which are not stored on the dApp, but instead stored on the wallet app.
+
+Thus, interaction between a dApp and a wallet is required such that:
+1) the dApp builds a transaction and makes it available for the wallet application
+2) wallet app shows a confirmation screen to the user, displaying the inboxes and 
+outboxes it is going to sign
+3) when user confirms, wallet app signs the transaction and submits it to the blockchain or
+returns back to the dApp
+4) dApp monitors the transaction on the blockchain and upon confirmations
+proceeds with its business logic.
+
+One possible implementation of such an interaction scheme is described in
+[EIP-0012](https://github.com/ergoplatform/eips/pull/23). It, however, only
+suitable for in-browser wallets, which can interact with dApps via javascript
+code injected into a web context.
+
+Luckily, the design of Ergo contracts allows for a simple and universal
+implementation suitable for mobile or desktop wallets. This EIP describes the interaction
+protocol and the reference implementation.
+
+In this 3 step process the main question is: 
+"How the transaction built in dApp can be transferred to the wallet
+application so that it can be signed there?".
+
+## ErgoPay Interaction Protocol
+
+An ErgoPay interaction between Wallet, and dApp is driven by dApp's
+user and proceeds as the following:
+
+1) The user enters the necessary information in the dApp's UI and proceeds to a
+payment screen. Optionally, dApp can request user's P2PK address with an extra 
+step using a signing request with a placeholder URL (see below).
+
+2) The payment screen shows transaction details and a QR code as well as a clickable 
+link.
+
+3) User clicks the link to hand the information over to a wallet application on the 
+same device, or scans the QR code using a wallet application.
+
+4) The Hot Wallet application parses the QR code data and obtains either
+`ErgoPaySigningRequest` or `ErgoPaySigningRequestUrl` data (see [Data Formats](#data-formats) section). In the later case the Hot
+Wallet uses the scanned url to download `ErgoPaySigningRequest` from the dApp.
+
+5) When `ErgoPaySigningRequest` is obtained, it is shown as a payment screen on the
```

Comment:
I think there's a 5a and 5b here based on `option 1` and `option 2` below - that is to say for the serverless `option 2` below it would be good to clarify here you don't need to create the entire `ErgoPaySigningRequest`, rather just the url safe ReducedTx

Reply:
Clarified